pkl-core/src/main/java/org/pkl/core/runtime/VmListingOrMapping.java

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.pkl.core.runtime;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.frame.MaterializedFrame;
+import com.oracle.truffle.api.nodes.IndirectCallNode;
+import org.graalvm.collections.EconomicMap;
+import org.graalvm.collections.UnmodifiableEconomicMap;
+import org.pkl.core.PklBugException;
+import org.pkl.core.ast.member.ListingOrMappingTypeCheckNode;
+import org.pkl.core.ast.member.ObjectMember;
+import org.pkl.core.util.EconomicMaps;
+import org.pkl.core.util.Nullable;
+
+public abstract class VmListingOrMapping<SELF extends VmListingOrMapping<SELF>> extends VmObject {
+
+  /** The original that this listing/mapping is a surrogate of. It might have its own surrogatee. */
+  private final @Nullable SELF surrogatee;
+
+  private final ListingOrMappingTypeCheckNode typeCheckNode;
+  private final MaterializedFrame typeNodeFrame;
+  private final EconomicMap<Object, ObjectMember> cachedMembers = EconomicMaps.create();
```

Comment:
I'm a little worried about the memory load this imposes (2x this for all `Listing`s and `Mapping`s; especially the `cachedMembers`). Have you measured the cost of this on a significant corpus?

Reply:
This only introduces new cached members on _some_ mappings and listings.

For example, this still contains 2 cached members, because `foo` is hidden and never executed:

```pkl
hidden foo: Mapping = new {
  ["first"] = 1
  ["second"] = 2
}

bar: Mapping<String, Int> = foo
```

Same here:

```pkl
hidden foo: Mapping<String, Int> = new {
  ["first"] = 1
}
```

However, this example means we get four cached members, because `foo` is not hidden:

```pkl
foo: Mapping = new {
  ["first"] = 1
  ["second"] = 2
}

bar: Mapping<String, Int> = foo
```

We can certainly profile this on some real-world projects to see what this change looks like.

Reply:
I added an additional optimization here: now, we will only store cached members on the delegated object _if_ the typecast results in a new object.

This results in two cached members:

```
foo: Mapping = new {
  ["first"] = 1
  ["second"] = 2
}

bar: Mapping<String, Int> = foo
```

This results in four cached members, due to the `Listing` to `Listing<Int>` typecast:

```pkl
foo: Mapping<Listing> = new {
  ["first"] = new { 1 }
  ["second"] = new { 2 }
}

bar: Mapping<String, Listing<Int>> = foo
```

Reply:
I also added another optimization: A typecast will only return a new delegating object _if_ the types are different.

This results in no delegation:

```pkl
foo: Listing<String>

bar = foo as Listing<String> // no delegation here because `foo` is already `Listing<String>`
```